pkg/generic/gwdeployer.go

```diff
@@ -0,0 +1,105 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generic
+
+import (
+	"fmt"
+
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	"github.com/submariner-io/cloud-prepare/pkg/k8s"
+	v1 "k8s.io/api/core/v1"
+)
+
+type gatewayDeployer struct {
+	k8sClient k8s.K8sInterface
+}
+
+// NewGatewayDeployer created a GatewayDeployer capable of deploying gateways to GCP
+func NewGatewayDeployer(k8sClient k8s.K8sInterface) (*gatewayDeployer, error) {
+	return &gatewayDeployer{k8sClient: k8sClient}, nil
+}
+
+func (g *gatewayDeployer) Deploy(input api.GatewayDeployInput, reporter api.Reporter) error {
+	gwNodes, err := g.k8sClient.ListGatewayNodes()
+	if err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	// Currently, we only support increasing the number of Gateway nodes which could be a valid use-case
+	// to convert a non-HA deployment to an HA deployment. We are not supporting decreasing the Gateway
+	// nodes (for now) as it might impact the datapath if we accidentally delete the active GW node.
+	if len(gwNodes.Items) < input.Gateways {
+		gatewayNodesToDeploy := input.Gateways - len(gwNodes.Items)
+
+		nonGWNodes, err := g.k8sClient.ListNodesWithLabel("!submariner.io/gateway")
```

Comment:
The add-on also searches for the `node-role.kubernetes.io/worker` label. I assume b/c you check for master node below that this isn't necessary?

Reply:
Yes. Also JFYI, in some k8s clusters like KIND, the worker nodes are not explicitly labelled with `node-role.kubernetes.io/worker`. Hence I'm using this approach of looking for master nodes with taints that would be generic.